### PR TITLE
[GLUTEN-7068][CORE] Fix issue updating leaf input metrics

### PR DESCRIPTION
**Pull request by ivoson**
## What changes were proposed in this pull request?

Fix issue #7068  collecting input metrics such as bytesRead and recordsRead from file scan.

## How was this patch tested?

UT added.

### CODE REVIEW

**Comment by ivoson** [comment]
cc @zhztheplayer @ulysses-you can you please review? Thanks

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala [hunk 1]
```diff
@@ -354,14 +356,30 @@ case class WholeStageTransformer(child: SparkPlan, materializeInput: Boolean = f
     }
   }
 
-  private def leafMetricsUpdater(): MetricsUpdater = {
-    child
-      .find {
-        case t: TransformSupport if t.children.forall(!_.isInstanceOf[TransformSupport]) => true
-        case _ => false
+  private def leafInputMetricsUpdater(): InputMetricsWrapper => Unit = {
+    def collectTransformSupportLeaves(
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala:360]: nit: Could rename to `collectLeaves`? To simplify code.

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala [hunk 2]
```diff
@@ -354,14 +356,30 @@ case class WholeStageTransformer(child: SparkPlan, materializeInput: Boolean = f
...
+      plan match {
+        case node: TransformSupport =>
+          if (plan.children.forall(!_.isInstanceOf[TransformSupport])) {
+            buffer.append(node)
+          } else {
+            plan.children
+              .filter(_.isInstanceOf[TransformSupport])
+              .foreach(collectTransformSupportLeaves(_, buffer))
+          }
+        case _ =>
```
Comment (zhztheplayer) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala:372]: nit:

```suggestion
        case node: TransformSupport if node.children.forall(!_.isInstanceOf[TransformSupport]) =>
            buffer.append(node)
        case node: TransformSupport =>
            node.children
              .foreach(collectTransformSupportLeaves(_, buffer))
        case _ =>
```

**Review by zhztheplayer — CHANGES_REQUESTED** [review]
Thank you for catching. Just some nits.

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala (hunk 1 above)
Comment (ivoson) [reply gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala:360]: Sure, will update in next iteration.

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala [hunk 3]
```diff
@@ -354,14 +356,30 @@ case class WholeStageTransformer(child: SparkPlan, materializeInput: Boolean = f
...
       }
-      .map(_.asInstanceOf[TransformSupport].metricsUpdater())
-      .getOrElse(MetricsUpdater.None)
+    }
+
+    val leavesBuffer = new ArrayBuffer[TransformSupport]()
+    collectTransformSupportLeaves(child, leavesBuffer)
+    val leavesMetricsUpdater = leavesBuffer.map(_.metricsUpdater())
```
[the remark this answers is not in the document]
Comment (ivoson) [reply gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala]: Updated, thanks.

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala (hunk 1 above)
Comment (ivoson) [reply gluten-substrait/src/main/scala/org/apache/gluten/execution/WholeStageTransformer.scala]: done

**Review by zhztheplayer — APPROVED** [review]
Thanks!

**Pull request merged**